spring-kafka/src/main/java/org/springframework/kafka/support/micrometer/KafkaListenerObservation.java

```diff
@@ -89,18 +242,45 @@ public static class DefaultKafkaListenerObservationConvention implements KafkaLi
 
 		@Override
 		public KeyValues getLowCardinalityKeyValues(KafkaRecordReceiverContext context) {
-			return KeyValues.of(KafkaListenerObservation.ListenerLowCardinalityTags.LISTENER_ID.asString(),
-							context.getListenerId());
+
+			return KeyValues.of(
+					ListenerLowCardinalityTags.LISTENER_ID.withValue(context.getListenerId()),
+					ListenerLowCardinalityTags.MESSAGING_SYSTEM.withValue("kafka"),
+					ListenerLowCardinalityTags.MESSAGING_OPERATION.withValue("receive"),
+					ListenerLowCardinalityTags.MESSAGING_SOURCE_NAME.withValue(context.getSource()),
+					ListenerLowCardinalityTags.MESSAGING_SOURCE_KIND.withValue("topic"),
+					ListenerLowCardinalityTags.MESSAGING_CONSUMER_GROUP.withValue(context.getGroupId())
+			);
+		}
+
+		@Override
+		@NonNull
+		public KeyValues getHighCardinalityKeyValues(KafkaRecordReceiverContext context) {
+			String clientId = context.getClientId();
+			KeyValues keyValues = KeyValues.of(
+					ListenerHighCardinalityTags.MESSAGING_PARTITION.withValue(context.getPartition()),
+					ListenerHighCardinalityTags.MESSAGING_OFFSET.withValue(context.getOffset()),
+					ListenerHighCardinalityTags.MESSAGING_CONSUMER_ID.withValue(getConsumerId(context, clientId))
+			);
+
+			if (StringUtils.hasText(clientId)) {
+				keyValues = keyValues
+						.and(ListenerHighCardinalityTags.MESSAGING_CLIENT_ID.withValue(clientId));
+			}
+
+			return keyValues;
 		}
 
 		@Override
 		public String getContextualName(KafkaRecordReceiverContext context) {
 			return context.getSource() + " receive";
 		}
 
-		@Override
-		public String getName() {
-			return "spring.kafka.listener";
+		private String getConsumerId(KafkaRecordReceiverContext context, @Nullable String clientId) {
```

Comment:
Looks like this new method can be `static`.

Reply:
Good point. 👍